M RIZIQ SIRFATULLAH ALFARIZI

i.2211266

# Load Data

In [86]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("erqizhou/students-data-analysis")

print("Path to dataset files:", path)

Path to dataset files: /root/.cache/kagglehub/datasets/erqizhou/students-data-analysis/versions/1


In [87]:
import os
dataset_path = "/root/.cache/kagglehub/datasets/erqizhou/students-data-analysis/versions/1"
files = os.listdir(dataset_path)
print("file didalam dataset",files)

file didalam dataset ['Students data.csv']


In [88]:
import pandas as pd

data_file = os.path.join(dataset_path, "Students data.csv")
data = pd.read_csv(data_file)

#print
data

,ID,class,gender,race,GPA,Algebra,Calculus1,Calculus2,Statistics,Probability,Measure,Functional_analysis,from1,from2,from3,from4,y
0,1141,A,male,1,73.47,64,81,87,60,74,71,60,A,A,A,3,0
1,1142,A,female,1,71.22,57,50,51,51,55,62,61,B,A,A,2,0
2,1143,A,female,2,74.56,47,48,71,60,61,68,64,C,A,A,0,1
3,1144,A,female,1,72.89,46,72,38,60,29,54,51,D,A,A,0,0
4,1145,A,female,1,70.11,49,45,63,60,66,66,61,E,A,A,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
100,1241,A,female,1,88.34,87,83,92,98,93,86,90,M,B,A,0,1
101,1242,B,male,1,89.84,98,77,95,98,96,88,100,A,B,A,0,1
102,1243,B,male,1,88.82,83,80,91,98,93,95,71,T,B,A,0,2
103,1244,A,male,1,86.60,92,82,91,99,94,82,78,S,B,A,0,2


# <B> Principal Component Analysis</b>

Apa itu PCA?

PCA Merupakan metode statistik yang digunakan untuk <b> Mereduksi Dimensi </b> dengan cara mengubah data ke dalam komponen Utama (Orincipal Coponent) yang memiliki variansi terbesar. PCA digunakan untuk mengatasi masalah dimensi pada data yang memiliki banyak fitur, tanpa kehilangan terlalu banyak informasi penting

<b>Manfaat PCA: </b>

1. Reduksi Dimensi
2. Visualisasi data
3. Menguragi Overfitting (Menghapus Fitur kurang penting)
4. Mempercepat komputasi

# Perbedaan akurasi penggunaan 3 data dan 5 data direduksi menjadi 3 data menggunakan PCA

In [89]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105 entries, 0 to 104
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   105 non-null    int64  
 1   class                105 non-null    object 
 2   gender               105 non-null    object 
 3   race                 105 non-null    int64  
 4   GPA                  105 non-null    float64
 5   Algebra              105 non-null    int64  
 6   Calculus1            105 non-null    int64  
 7   Calculus2            105 non-null    int64  
 8   Statistics           105 non-null    int64  
 9   Probability          105 non-null    int64  
 10  Measure              105 non-null    int64  
 11  Functional_analysis  105 non-null    int64  
 12  from1                105 non-null    object 
 13  from2                105 non-null    object 
 14  from3                105 non-null    object 
 15  from4                105 non-null    int

In [90]:
data.columns

Index(['ID', 'class', 'gender', 'race', 'GPA', 'Algebra', 'Calculus1',
       'Calculus2', 'Statistics', 'Probability', 'Measure',
       'Functional_analysis', 'from1', 'from2', 'from3', 'from4', 'y'],
      dtype='object')

# Klasifikasi menggunakan Support Vector Machine (tanpa PCA)

In [91]:
x = data[['Algebra', 'Calculus1',]]
y = data['y']

In [92]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=0)

In [93]:
# Support Vector Classidication (SVC)
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

model_svc = SVC()
model_svc.fit(x_train, y_train)
y_pred = model_svc.predict(x_test)
akurasi = accuracy_score(y_test, y_pred)

print(f"Hasil Akurasi Menggunakan SVM = {akurasi:.2%}")

Hasil Akurasi Menggunakan SVM = 61.90%


# Menggukan PCA

In [94]:
x = data[['Algebra', 'Calculus1',
       'Calculus2', 'Statistics', 'Probability', 'Measure',
       'Functional_analysis']]
y = data['y']

In [95]:
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [96]:
# Preprocessing
# Normalisasi data (Data di-scaling agar memiliki varians yang seragam)
scaler = StandardScaler()
x_scaled = scaler.fit_transform(x)

In [97]:
pca = PCA(n_components=2) # mengurangi feature menjadi 2 feature
x_pca = pca.fit_transform(x_scaled)

In [98]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x_pca, y, test_size=0.2, random_state=0)

In [99]:
# Support Vector Classidication (SVC)
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

model_svc2 = SVC()
model_svc2.fit(x_train, y_train)
y_pred = model_svc2.predict(x_test)
akurasi = accuracy_score(y_test, y_pred)

print(f"Hasil Akurasi Menggunakan SVM = {akurasi:.2%}")

Hasil Akurasi Menggunakan SVM = 66.67%


Berdasarkan Hasil pertama, data yang digunkan untuk training hanya 2 feature saja memberikan akurasi sebesar 61.9%.

sedangkan data dengan 5 feature kemudian direduksi menjadi 2 fitur agar informasi data sebelumnya tidak hilang, memberikan hasil akurasi sebesar 66.7%

KESIMPULAN

Untuk dapat mempertahankan tanpa mengurangi akurasi yang besarm disarankan untuk mengambil fitur sebanyak-banyaknya dan melalukan reduksi menggunakan PCA jika tujuaanya untuk Visualisasi 2 dimensi (2 Feature) atau 3 dimensi (3 feature)

# Feature Importance & PCA

In [100]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105 entries, 0 to 104
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   105 non-null    int64  
 1   class                105 non-null    object 
 2   gender               105 non-null    object 
 3   race                 105 non-null    int64  
 4   GPA                  105 non-null    float64
 5   Algebra              105 non-null    int64  
 6   Calculus1            105 non-null    int64  
 7   Calculus2            105 non-null    int64  
 8   Statistics           105 non-null    int64  
 9   Probability          105 non-null    int64  
 10  Measure              105 non-null    int64  
 11  Functional_analysis  105 non-null    int64  
 12  from1                105 non-null    object 
 13  from2                105 non-null    object 
 14  from3                105 non-null    object 
 15  from4                105 non-null    int

In [101]:
from sklearn.preprocessing import LabelEncoder
label_encoder = LabelEncoder()
data['class_encode'] = label_encoder.fit_transform(data['class'])
data['gender_encode'] = label_encoder.fit_transform(data['gender'])
data['from1_encode'] = label_encoder.fit_transform(data['from1'])
data['from2_encode'] = label_encoder.fit_transform(data['from2'])
data['from3_encode'] = label_encoder.fit_transform(data['from3'])

In [102]:
# menghapus data kategori
data2 = data.drop(columns=['ID','class', 'gender', 'from1', 'from2', 'from3'])
data2

,race,GPA,Algebra,Calculus1,Calculus2,Statistics,Probability,Measure,Functional_analysis,from4,y,class_encode,gender_encode,from1_encode,from2_encode,from3_encode
0,1,73.47,64,81,87,60,74,71,60,3,0,0,1,0,0,0
1,1,71.22,57,50,51,51,55,62,61,2,0,0,0,3,0,0
2,2,74.56,47,48,71,60,61,68,64,0,1,0,0,4,0,0
3,1,72.89,46,72,38,60,29,54,51,0,0,0,0,5,0,0
4,1,70.11,49,45,63,60,66,66,61,0,0,0,0,6,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
100,1,88.34,87,83,92,98,93,86,90,0,1,0,0,14,1,0
101,1,89.84,98,77,95,98,96,88,100,0,1,1,1,0,1,0
102,1,88.82,83,80,91,98,93,95,71,0,2,1,1,21,1,0
103,1,86.60,92,82,91,99,94,82,78,0,2,0,1,20,1,0


In [103]:
data2.columns

Index(['race', 'GPA', 'Algebra', 'Calculus1', 'Calculus2', 'Statistics',
       'Probability', 'Measure', 'Functional_analysis', 'from4', 'y',
       'class_encode', 'gender_encode', 'from1_encode', 'from2_encode',
       'from3_encode'],
      dtype='object')

In [104]:
x = data2.drop(columns=['y'])
y = data2['y']

In [105]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=0)

In [106]:
# Support Vector Classidication (SVC)
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

model_svc = SVC(kernel = 'linear')
model_svc.fit(x_train, y_train)
y_pred = model_svc.predict(x_test)
akurasi = accuracy_score(y_test, y_pred)

print(f"Hasil Akurasi Menggunakan SVM = {akurasi:.2%}")

Hasil Akurasi Menggunakan SVM = 38.10%


In [112]:
import numpy as np
feature_importance = pd.DataFrame({
    'Feature': x.columns,
    'Importance': np.abs(model_svc.coef_[0])
})
feature_importance = feature_importance.sort_values(by='Importance', ascending=False)
feature_importance


,Feature,Importance
0,race,0.907236
11,gender_encode,0.660453
10,class_encode,0.609905
13,from2_encode,0.484200
9,from4,0.264502
14,from3_encode,0.254405
6,Probability,0.095069
7,Measure,0.080043
1,GPA,0.070262
12,from1_encode,0.065424


Berdasarkan Feature importance, terdapat 8 fitur yang nilainya kurang dari 0.1, Sehingga fitur tersebut tidak memberikan dampak besar pada model yg digunakan, Jadi perlu dilakukan training ulang dengan <b>mereduksi </b>dimensi menjadi 6 fitur, karena memberikan nilai lebih dari 0.1

In [117]:
x = data2.drop(columns=['y'])
y = data2['y']

In [118]:
# Preprocessing
# Normalisasi data (Data di-scaling agar memiliki varians yang seragam)
scaler = StandardScaler()
x_scaled = scaler.fit_transform(x)

In [119]:
pca = PCA(n_components=6) # mengurangi 14 feature menjadi 6 feature
x_pca = pca.fit_transform(x_scaled)

In [120]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x_pca, y, test_size=0.2, random_state=0)

In [122]:
# Support Vector Classidication (SVC)
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

model_svc2 = SVC(kernel = 'linear')
model_svc2.fit(x_train, y_train)
y_pred = model_svc2.predict(x_test)
akurasi = accuracy_score(y_test, y_pred)

print(f"Hasil Akurasi Menggunakan SVM = {akurasi:.2%}")

Hasil Akurasi Menggunakan SVM = 57.14%


Kesimpulan: <br>
Ternyata Pengurangan dimensi menjadi 6 fitur karena 8 fiturnya tidak memberikan pengaruh terhadap model, memberikan akurasi lebih tinggi, sehingga 8 fitur yang dibuang sebelumnya memberikan pengaruh akurasi lebih rendah. (The End)